packages/dhis2/src/Adaptor.js

```diff
@@ -508,33 +565,52 @@ export function upsert(
   return state => {
     console.log(`Preparing upsert via 'get' then 'create' OR 'update'...`);
 
-    // NOTE: that these parameters are all expanded by the `get`, `create`, and
-    // `update` functions used inside this composed "upsert" function.
-    return get(
-      resourceType,
-      query,
-      options
-    )(state)
-      .then(resp => {
-        const resources = resp.data[resourceType];
-        if (resources.length > 1) {
-          throw new RangeError(
-            `Cannot upsert on Non-unique attribute. The operation found more than one records for your request.`
-          );
-        } else if (resources.length <= 0) {
-          return create(resourceType, data, options)(state);
-        } else {
-          // Pick out the first (and only) resource in the array and grab its
-          // ID to be used in the subsequent `update` by the path determined
-          // by the `selectId(...)` function.
-          const path = resources[0][selectId(resourceType)];
-          return update(resourceType, path, data, options)(state);
-        }
-      })
-      .then(result => {
-        console.log(`Performed a "composed upsert" on ${resourceType}`);
+    const resolvedResourceType = expandReferences(resourceType)(state);
+    const resolvedOptions = expandReferences(options)(state);
+    const resolvedData = expandReferences(data)(state);
+    if (shouldUseNewTracker(resolvedResourceType)) {
+      const { params, requestConfig } = resolvedOptions;
+      const { configuration } = state;
+      return callNewTracker(
+        'create',
+        configuration,
+        { importStrategy: 'CREATE_AND_UPDATE' },
+        nestArray(resolvedData, resolvedResourceType),
+        params,
+        requestConfig
+      ).then(result => {
+        console.log(`Performed a "composed upsert" on ${resolvedResourceType}`);
```

Comment:
Since the `.then()` is the same, let's re-use the promise, like we do with create